addons/addon-base-raas/packages/base-raas-services/lib/study/study-operation-service.js

```diff
@@ -0,0 +1,286 @@
+const { getSystemRequestContext } = require('@aws-ee/base-services/lib/helpers/system-context');
+const { processInBatches } = require('@aws-ee/base-services/lib/helpers/utils');
+
+const { StudyPolicy } = require('../helpers/iam/study-policy');
+const { hasAccess, accessLevels } = require('./helpers/entities/study-methods');
+
+const extensionPoint = 'study-access-strategy';
+
+// An impacted user is a user who is listed in either usersToAdd section and/or usersToRemove section of
+// the updateRequest object.
+const getImpactedUsers = (updateRequest = {}) => {
+  const addedUsers = _.map(updateRequest.usersToAdd, item => item.uid);
+  const removedUsers = _.map(updateRequest.usersToRemove, item => item.uid);
+
+  return _.uniq([...addedUsers, ...removedUsers]);
+};
+
+/**
+ * This service is a higher level service than the study service. It understands that studies are mounted on
+ * workspaces, and it orchestrates all the necessary logic when an operation is needed to be performed on studies.
+ * For example, it orchestrates all the necessary logic when study permission changes and propagates all the way to
+ * the relevant workspaces.
+ */
+class StudyOperationService extends Service {
+  constructor() {
+    super();
+    this.dependency([
+      'auditWriterService',
+      'lockService',
+      'studyService',
+      'environmentScService',
+      'pluginRegistryService',
+    ]);
+  }
+
+  /**
+   * This method not only updates the study/study permissions entities, but it also understands how to propagate the
+   * permission changes to the workspaces. Therefore, this service is a higher level service than the study service.
+   * It understands that studies are mounted on workspaces, and it orchestrates all the necessary logic to make
+   * study permission changes propagate all the way to the relevant workspaces.
+   *
+   * @param requestContext The standard requestContext
+   * @param studyId The study id
+   * @param updateRequest The updateRequest object containing the information about which users to add/remove
+   * including permission levels for the given study entity.
+   *
+   * The updateRequest should have the following shape:
+   * {
+   *  usersToAdd: [ {uid, permissionLevel}, ...]
+   *  usersToRemove: [ {uid, permissionLevel}, ...]
+   * }
+   *
+   * @returns studyPermissionsEntity
+   */
+  async updatePermissions(requestContext, studyId, updateRequest) {
+    const [studyService, lockService, environmentScService] = await this.service([
+      'studyService',
+      'lockService',
+      'environmentScService',
+    ]);
+
+    // First, we need to have a lock at the study operation level
+    const result = await lockService.tryWriteLockAndRun({ id: `study-${studyId}-operation` }, async () => {
+      // Currently, propagating permission is not done via a workflow, so we have to do everything in less than
+      // 30 seconds. In an attempt to fail fast, an exception is thrown if we are trying to update more than
```

Comment:
Do we need a special note or spinning on the UI so user don't see it as a bug when it's taking a lot of time?

Reply:
The existing UI for updating study permissions has the spinner implemented, so we are good here.